### PR TITLE
Implement Partial Replan in RRTPlanner

### PR DESCRIPTION
**Pull request by ashaw596**

### CODE REVIEW

File: soccer/planning/RRTPlanner.cpp [hunk 1]
```diff
@@ -17,10 +19,60 @@ using namespace Eigen;
 using namespace Geometry2d;
 
 namespace Planning {
+REGISTER_CONFIGURABLE(RRTPlanner);
+
+ConfigDouble* RRTPlanner::_partialReplanLeadTime;
+
+void RRTPlanner::createConfiguration(Configuration* cfg) {
+    _partialReplanLeadTime = new ConfigDouble(
+        cfg, "RRTPlanner/partialReplanLeadTime", 0.2, "partialReplanLeadTime");
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: You might as well leave the "description" parameter out if it's the same as the name.

File: soccer/planning/RRTPlanner.cpp [hunk 2]
```diff
@@ -17,10 +19,60 @@ using namespace Eigen;
...
+
+    if (*motionConstraints._replan_threshold != 0 &&
+        pathError > replanThreshold) {
+        return true;
+    }
+
+    return false;
+}
+
+boost::optional<float> shouldPartialReplan(
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: Could this method be renamed for clarity?  The word "should" usually means it's a method that returns a boolean.

File: soccer/planning/RRTPlanner.cpp [hunk 3]
```diff
@@ -55,7 +107,28 @@ bool RRTPlanner::shouldReplan(const SinglePlanRequest& planRequest,
...
+        debugLog("PathPlanning Failed");
+        return nullptr;
+    }
+
+    // Generate and return a cubic bezier path using the waypoints
+    return generateCubicBezier(points, obstacles, motionConstraints, start.vel,
+                               goal.vel);
+}
+
+const int maxContinue = 60;
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: What's `maxContinue`?

File: soccer/planning/RRTPlanner.cpp [hunk 4]
```diff
@@ -17,10 +19,60 @@ using namespace Eigen;
...
+        goalVelDiff > RRTPlanner::goalChangeThreshold()) {
+        // FIXME: goalChangeThreshold shouldn't be used for velocities as it
+        // is above
+        return prevPath->getDuration() - timeIntoPath;
+    }
+
+    return boost::none;
+}
+
 bool RRTPlanner::shouldReplan(const SinglePlanRequest& planRequest,
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: Is this method used anymore?  If not, let's delete it.

File: soccer/planning/RRTPlanner.cpp [hunk 5]
```diff
@@ -55,7 +107,28 @@ bool RRTPlanner::shouldReplan(const SinglePlanRequest& planRequest,
     return false;
 }
 
-const int maxContinue = 10;
+std::unique_ptr<Path> RRTPlanner::plan(
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: Where is  `plan()` called from and how does it differ from `run()`?

File: soccer/planning/RRTPlanner.cpp [hunk 6]
```diff
@@ -89,32 +162,99 @@ std::unique_ptr<Path> RRTPlanner::run(SinglePlanRequest& planRequest) {
...
         }
-        return std::move(path);
     } else {
-        if (reusePathTries >= maxContinue) {
+        fullReplan = true;
+        prevPathInval = true;
+    }
+
+    if (fullReplan || prevPathInval) {
+        printf("shouldFullReplan\n");
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: x

File: soccer/planning/Path.hpp [hunk 7]
```diff
@@ -229,6 +229,82 @@ class AngleFunctionPath : public Path {
     virtual void setStartTime(RJ::Time t) override { path->setStartTime(t); }
 };
 
+class EmptyPath : public Path {
```
Comment (justbuchanan) [inline soccer/planning/Path.hpp]: What's the use case for `EmptyPath`?

File: soccer/planning/RRTPlanner.cpp [hunk 8]
```diff
@@ -55,7 +107,28 @@ bool RRTPlanner::shouldReplan(const SinglePlanRequest& planRequest,
...
+    }
+
+    // Generate and return a cubic bezier path using the waypoints
+    return generateCubicBezier(points, obstacles, motionConstraints, start.vel,
+                               goal.vel);
+}
+
+const int maxContinue = 60;
 
 std::unique_ptr<Path> RRTPlanner::run(SinglePlanRequest& planRequest) {
```
Comment (justbuchanan) [inline soccer/planning/RRTPlanner.cpp]: This method could use a few comments. It's a little tough to follow what all's going on in all the if-statements and things.

**Comment by ashaw596** [comment]
Hmm, its the RRT is is sort of cause some weird giant paths to be planned compared to the old planner. Like this https://www.dropbox.com/s/nhtkrdz7j1z1wha/Screenshot%202016-11-05%2003.31.01.png?dl=0

**Comment by ashaw596** [comment]
On hiatus. Probably gonna redo in different pull request.

**Pull request closed without merging**